Mathlib/Tactic/Ring.lean

```diff
@@ -0,0 +1,490 @@
+  | _ => do
+    let e₂ ← mkRawNatLit (m.2 - 1)
+    let (tl, hl) ← evalPow he (e₂, m.2-1)
+    let (t, p₂) ← evalMul tl he
+    return (t, ← mkAppM ``pow_succ_eq #[e, e₂, tl, t, hl, p₂])
+
+
+theorem horner_atom {α} [CommSemiring α] (x : α) : x = horner 1 x 1 0 := by
+  simp [horner]
+
+/-- Evaluate `a` where `a` is an atom. -/
+def evalAtom (e : Expr) : RingM (HornerExpr × Expr) := do
+  let i ← addAtom e
+  let zero ← const (← mkAppOptM ``OfNat.ofNat #[(← read).α, mkRawNatLit 0, none]) 0
+  let one ← const (← mkAppOptM ``OfNat.ofNat #[(← read).α, mkRawNatLit 1, none]) 1
+  return (← xadd' one (e,i) (mkRawNatLit 1,1) zero, ← mkAppM ``horner_atom #[e])
+
+
+theorem subst_into_add {α} [Add α] (l r tl tr t)
+  (prl : (l : α) = tl) (prr : r = tr) (prt : tl + tr = t) : l + r = t :=
+by rw [prl, prr, prt]
+
+theorem subst_into_mul {α} [Mul α] (l r tl tr t)
+  (prl : (l : α) = tl) (prr : r = tr) (prt : tl * tr = t) : l * r = t :=
+by rw [prl, prr, prt]
+
+theorem subst_into_pow {α} [Monoid α] (l r tl tr t)
+  (prl : (l : α) = tl) (prr : (r : ℕ) = tr) (prt : tl ^ tr = t) : l ^ r = t :=
+by rw [prl, prr, prt]
+
+partial def eval (e : Expr) : RingM (HornerExpr × Expr) :=
+  match e.getAppFnAndArgs with
+  | some (``HAdd.hAdd, #[_,_,_,_,e₁,e₂]) => do
+    let (e₁', p₁) ← eval e₁
+    let (e₂', p₂) ← eval e₂
+    let (e', p') ← evalAdd e₁' e₂'
+    let p ← mkAppM ``subst_into_add #[e₁, e₂, e₁', e₂', e', p₁, p₂, p']
+    (e',p)
+  | some (``HMul.hMul, #[_,_,_,_,e₁,e₂]) => do
+    let (e₁', p₁) ← eval e₁
+    let (e₂', p₂) ← eval e₂
+    let (e', p') ← evalMul e₁' e₂'
+    let p ← mkAppM ``subst_into_mul #[e₁, e₂, e₁', e₂', e', p₁, p₂, p']
+    return (e', p)
+  | some (``HPow.hPow, #[_,_,_,P,e₁,e₂]) => do
+    -- let (e₂', p₂) ← lift $ norm_num.derive e₂ <|> refl_conv e₂,
+    let (e₂', p₂) ← (e₂, ← mkEqRefl e₂)
+    match e₂'.numeral?, P.getAppFn with
+    | some k, Expr.const ``Monoid.HPow _ _ => do
+      let (e₁', p₁) ← eval e₁
+      let (e', p') ← evalPow e₁' (e₂, k)
+      let p ← mkAppM ``subst_into_pow #[e₁, e₂, e₁', e₂', e', p₁, p₂, p']
+      return (e', p)
+    | _, _ => do ← evalAtom e
+    evalAtom e
+  | _ =>
+    match e.numeral? with
+    | some n => (const e n).reflConv
+    | _ => (evalAtom e)
+
+
+elab "ring" : tactic => do
+  let g ← getMainTarget
+  match g.getAppFnAndArgs with
+  | some (`Eq, #[ty, e₁, e₂]) =>
+    let ((e₁', p₁), (e₂', p₂)) ← run e₁ $ Prod.mk <$> eval e₁ <*> eval e₂
+    if (← isDefEq e₁' e₂') then
+      let p ← mkEqTrans p₁ (← mkEqSymm p₂)
+      ensureHasNoMVars p
+      assignExprMVar (← getMainGoal) p
+
+      replaceMainGoal []
+    else
+      throwError "failed \n{← e₁'.pp}\n{← e₂'.pp}"
+  | _ => throwError "failed: not an equality"
+
+example (x y : ℕ) : x + y = y + x := by ring
+example (x y : ℕ) : x + y + y = 2 * y + x := by ring
+example (x y : ℕ) : x + id y = y + id x := by ring
+-- example {α} [CommRing α] (x y : α) : x + y + y - x = 2 * y := by ring
```

Comment:
Some of these examples are commented out.  I assume that's because they don't work at the moment (but did in Lean 3)?  Is this because we don't have the type classes / types yet in mathlib4?  Or is there something else that's missing?

Reply:
All of them worked in Lean 3 yes (even though some used `ring_nf`). The missing features for them to work here are:

- Compatibility of `ring` with subtraction and division
- Defining ℚ, ℝ, fields and linear ordered fields